tests/core/kernel-base-gadget-pair-single-reboot/task.yaml

```diff
@@ -18,53 +19,101 @@ environment:
     SNAP_ONE/kernel_gadget: pc
     SNAP_TWO/kernel_gadget: pc-kernel
 
-    BLOB_DIR: $(pwd)/fake-store-blobdir
+    # We are uploading too much in terms of size for the snaps
+    # we sideload. On UC18 this is disabled automatically, but
+    # on UC20+ this has proved to be an issue
+    SNAPD_NO_MEMORY_LIMIT: 1
 
 prepare: |
-    if [ "$TRUST_TEST_KEYS" = "false" ]; then
-        tests.exec skip-test "This test needs test keys to be trusted" && exit 0
-    fi
-    if os.query is-core18 && [[ "$SPREAD_BACKEND" =~ openstack ]]; then
-        tests.exec skip-test "The kernel snap used for openstack cannot be refreshed with Canonical model" && exit 0
-    fi
+    # To get VERSION_ID defined
+    . /etc/os-release
 
-    core_snap=core20
-    if os.query is-core18; then
-        core_snap=core18
-    fi
+    # does not work for 16, but we don't support 16 for this
+    core_snap="core${VERSION_ID}"
 
     # read all for simplification
     readlink /snap/pc-kernel/current > pc-kernel.rev
     readlink "/snap/$core_snap/current" > core.rev
-    readlink "/snap/pc/current" > pc.rev
-
-restore: |
-    tests.exec is-skipped && exit 0
+    readlink /snap/pc/current > pc.rev
 
-    "$TESTSTOOLS"/store-state teardown-fake-store "$BLOB_DIR"
+debug: |
+    change_id="$(cat refresh-change-id)"
+    snap debug state --change "$change_id" /var/lib/snapd/state.json || true
 
 execute: |
-    tests.exec is-skipped && exit 0
+    # To get VERSION_ID defined
+    . /etc/os-release
 
-    core_snap=core20
-    if os.query is-core18; then
-        core_snap=core18
+    if snap list pc | NOMATCH "x[1-9]"; then
+        echo "This test needs a host using an unasserted 'pc' gadget snap"
+        exit 1
     fi
 
+    # does not work for 16, but we don't support 16 for this
+    core_snap="core${VERSION_ID}"
+    
+    # correct the naming based on the system we are testing
     if [ "$SNAP_ONE" = "core" ]; then
         SNAP_ONE=$core_snap
     elif [ "$SNAP_TWO" = "core" ]; then
         SNAP_TWO=$core_snap
     fi
 
+    wait_and_verify_change() {
+        local CHANGE_ID="$1"
+        snap watch "$CHANGE_ID" || true
+        snap changes | MATCH "$CHANGE_ID\s+Done"
+        snap change "$CHANGE_ID" > tasks.done
+        # nothing has failed
+        NOMATCH '^Error' < tasks.done
+        # nothing was undone
+        NOMATCH '^Undone' < tasks.done
+        # nothing is waiting
+        NOMATCH '^Wait' < tasks.done
+
+        # we did not even try to hijack shutdown (/bin/systemctl) because that
+        # could race with snapd (if that wanted to call it), so just check that
+        # the system is in a stable state once we have already determined that
+        # the change is complete
+        # XXX systemctl exits with non-0 when in degraded state
+        # Note: on bionic, is-system-running does not support --wait
+        retry -n 30 sh -c '(systemctl is-system-running || true) | MATCH "(running|degraded)"'
```

Comment:
Do we need the `|| true` part? With a pipeline, the exit code of the left-hand side element is not relevant.